File: java/src/patterns/creational/factory/FactoryClient.java
package patterns.creational.factory;

import patterns.creational.factory.abstractcreator.ParserCreator;
import patterns.creational.factory.concretecreators.CsvParserCreator;
import patterns.creational.factory.concretecreators.JsonParserCreator;
import patterns.creational.factory.concretecreators.XmlParserCreator;

public class FactoryClient {
    public static void main(String[] args) {
        ParserCreator creator;

        creator = new CsvParserCreator();
        creator.parse("data.csv");

        creator = new JsonParserCreator();
        creator.parse("data.json");

        creator = new XmlParserCreator();
        creator.parse("data.xml");
    }
}

/*
Adding a new parser is simple:
Example:

class HtmlParser implements DocumentParser {
    @Override
    public void parse(String document) {
        System.out.println("Parsing HTML document: " + document);
    }
}

class HtmlParserCreator extends ParserCreator {
    @Override
    public DocumentParser createParser() {
        return new HtmlParser();
    }
}

creator = new HtmlParserCreator();
creator.parse("data.html");

*/

File: python/src/patterns/creational/factory/factory_client.py
from src.patterns.creational.factory.concrete_creators.csv_parser_creator import CsvParserCreator
from src.patterns.creational.factory.concrete_creators.json_parser_creator import JsonParserCreator
from src.patterns.creational.factory.concrete_creators.xml_parser_creator import XmlParserCreator

def main():
    csv_parser_creator = CsvParserCreator()
    json_parser_creator = JsonParserCreator()
    xml_parser_creator = XmlParserCreator()

    csv_parser_creator.parse("data.csv")
    json_parser_creator.parse("data.json")
    xml_parser_creator.parse("data.xml")

if __name__ == "__main__":
    main()

"""
Adding a new parser is simple:
Example:

class HtmlParser(DocumentParser):
    def parse(self, document: str) -> None:
        print(f"Parsing HTML document: {document}")

class HtmlParserCreator(ParserCreator):
    def create_parser(self) -> DocumentParser:
        return HtmlParser()

html_parser_creator = HtmlParserCreator()
html_parser_creator.parse("data.html")
"""
